### DISCUSSION TITLE
JellySearch

### DISCUSSION THREAD

**Discussion by Chaphasilor** [UnicornsOnLSD/finamp#969] [General]
Hey @DomiStyle!
I just stumbled upon your JellySearch project and it looks really promising!  
Search has been a major issue of Jellyfin for a while now, and your solution seems nice.

If I understood things correctly, this should work out-of-the-box with Finamp, right?  

Also, there have been some issues about improving the search in Finamp, like #400, #257, #451 and #379. Do you think it would be possible to somehow optionally integrate with JellySearch to fix some of these? Specifically better sorting and more deliberate fuzzy matching would be nice!

Sorry about the weird way of reaching out ^^

**Comment by Nerokor** [comment]
Hey, I'm a user of [Jellysearch](https://gitlab.com/DomiStyle/jellysearch), and it's the best improvement you can really feel in Jellyfin in a while. It works with Finamp, and I think it solves a few of the issues you've linked. For example, searching for "rage machine" indeed provides the artist Rage Against the Machine.

I highly recommend setting it up on your server to test it out. I've contacted DomiStyle early on with issues, and he's always responded right away on GitLab.

While I'm here...

One request I have is for a universal search feature in Finamp. Now that the search functionality is much faster, it has become inconvenient to switch between tabs—such as "artist," "album," and "track"—to get the results I want for those specific results. It would be much more convenient if there were a single search bar that displayed categorical results, similar to how [Feishin](https://github.com/jeffvli/feishin) (a desktop client for Jellyfin/Subsonic) operates.

Additionally, I noticed that the search became a bit slower after a change made in the last beta. It seems that searches are now delayed a bit with each character typed if I recall? This change has made the search experience slower, while using Jellysearch.

**Reply by Zedniac** [reply]
I can confirm that Jellysearch works great with Finamp. If it didn't, I almost certainly wouldn't be using Finamp (through no fault of Finamp's, but rather Jellyfin's incredibly slow search on large libraries). Took my queries from like 15 seconds to 0-2 seconds. Running it in Docker alongside Jellyfin.

**Reply by Chaphasilor** [reply]
@Nerokor universal search is definitely planned, and much of the plumbing for it is already in place. It will probably drop along with adding a proper home screen, but might take a while because I'm super busy!

And you're right, the search is now being "debounced", essentially waiting until the user finishes typing before sending a request to the server. Without this, we would spam the server with requests, which makes the regular even slower. So this change results in a noticable speedup for regular Jellyfin setups.  
It's a bummer that this actually makes search a bit slower in your case, but it speaks to the quality/performance of JellySearch. The delay should only be minor though, the timeout is just 400 ms. Once we "officially" support JellySearch, we can add a toggle or automatic detection that disables the debouncing to speed things up even more.

**Comment by DomiStyle** [comment]
> If I understood things correctly, this should work out-of-the-box with Finamp, right?

Yes, have been using it for a few months with Finamp and it works perfectly fine. Search was pretty much broken since 10.9 otherwise.

> Do you think it would be possible to somehow optionally integrate with JellySearch to fix some of these?

All of these should be fixed out of the box with JellySearch.

> Specifically better sorting and more deliberate fuzzy matching would be nice!

Fuzzy matching works pretty well with Meilisearch.

When searching any other sorting parameter is ignored and the order in which Meilisearch returns the results is always returned as is.

When not searching, Jellyfin handles any sorting and JellySearch is never involved.

**Reply by Chaphasilor** [reply]
Awesome, that sounds great!  
A few more questions...

> All of these should be fixed out of the box with JellySearch.
For #379 in particular, how much control do you have about the sort order / prioritization of results? Is it just a default behavior, or is it customizable?

Would it theoretically be possible to apply any additional sorting parameters to the Meilisearch results? Or will sorting never work, even if you try to support it in the proxy?  

For the redesign we'll probably add more sophisticated sorting and filtering options, and I'm not sure how (if) that will be integrated with the search. Just asking to check if there are any roadblocks to supporting/recommending your proxy :)

**Reply by DomiStyle** [reply]
> Would it theoretically be possible to apply any additional sorting parameters to the Meilisearch results? Or will sorting never work, even if you try to support it in the proxy?

Yes, Meilisearch does support filtering and sorting. Currently there's only some very light sorting going on to nudge Meilisearch to prefer items with higher critic/community ratings: https://gitlab.com/DomiStyle/jellysearch/-/blob/main/src/JellySearch/Jobs/IndexJob.cs?ref_type=heads#L45

Real filtering/sorting while searching is probably a niche use case though.

> Just asking to check if there are any roadblocks to supporting/recommending your proxy :)

There's also another implementation now for people without reverse proxy: https://github.com/arnesacnussem/jellyfin-plugin-meilisearch

JellySearch or the plugin above should probably do the job until Jellyfin supports search plugins properly.
